modules/service/src/main/scala/lucuma/odb/service/GuideEnvironmentService.scala

```diff
@@ -0,0 +1,575 @@
+                         obsInfo.agsParams,
+                         candidates.toList
+            )
+            .sortUsablePositions
+          NonEmptyList
+            .fromList(analysis)
+            .toRight(Error.GeneralError(s"No usable guide targets were found for observation ${obsInfo.id}"))
+        }
+
+      override def get(pid: Program.Id, oid: Observation.Id, obsTime: Instant)(using
+        NoTransaction[F]
+      ): F[Either[Error, GuideEnvironment]] =
+        (for {
+          obsInfo        <- EitherT(getObservationInfo(pid, oid))
+          asterism       <- EitherT(getAsterism(pid, oid))
+          tracking        = ObjectTracking.fromAsterism(asterism)
+          so             <- EitherT(getOffsets(pid, oid))
+          (site, offsets) = so
+          candidates     <- EitherT(getCandidates(oid, obsTime, tracking))
+          baseCoords     <- EitherT.fromEither(
+                              obsInfo.explicitBase
+                                .orElse(
+                                  tracking
+                                    .at(obsTime)
+                                    .map(_.value)
+                                )
+                                .toRight(
+                                  Error.GeneralError(
+                                    s"Unable to get coordinates for asterism in observation $oid"
+                                  )
+                                )
+                            )
+          scienceCoords  <- EitherT.fromEither(
+                              asterism.toList
+                                .traverse(t => ObjectTracking.fromTarget(t).at(obsTime).map(_.value))
+                                .toRight(
+                                  Error.GeneralError(
+                                    s"Unable to get coordinates for science targets in observation $oid"
+                                  )
+                                )
+                            )
+          positions      <- EitherT.fromEither(
+                              getPositions(oid, site, obsInfo.posAngleConstraint, offsets, tracking, obsTime)
+                            )
+          usable         <- EitherT.fromEither(
+                              processCandidates(obsInfo, baseCoords, scienceCoords, positions, candidates)
+                            )
+          target          = GuideStarCandidate.siderealTarget.reverseGet(usable.head.target)
+          posAngle       <-
+            // "Can't" fail because all `Usable` candidates have a position angle
+            EitherT.fromEither(usable.head.posAngle.toRight(Error.GeneralError("No position angle for guide target.")))
+          ge              = GuideEnvironment(posAngle, List(GuideTarget(guideProbe, target)))
+        } yield ge).value
+    }
+
+  object Statements {
+    import ProgramService.Statements.andWhereUserAccess
+
+    def getAsterism(user: User, pid: Program.Id, oid: Observation.Id): AppliedFragment =
```

Comment:
Maybe we should move this to the `AsterismService`?

Reply:
I was torn about that, since the `AsterismService` doesn't care about being able to actually get an in-memory Target or Asterism. It's methods are only concerned with mutations, not queries. I am open to the idea of moving it, but am not convinced it (and the associated skunk decoder) belong there.

Reply:
`AsterismService` doesn't do queries or get an in-memory representation because it didn't need to before now.  I suggested it because if I needed to query for asterisms in the future the first place I'd probably look is in the `AsterismService` and not in the `GuideEnvironmentService`.

Reply:
Good point. You've convinced me. I'll move them.